In [1]:
# Importing the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# read the csv file
data = pd.read_csv('../data/discipline_data_cumberland.csv')
print(f'Shape of dataset: {data.shape}\n')
data.head()

Shape of dataset: (311942, 17)



STUDENTID  INCIDENT_ID INCIDENT_CATEGORY OFFENSE_DATE  SCHOOL_NUMBER  \
0     143624       205050          Offender   2017-02-21       260322.0   
1     143624       205050          Offender   2017-02-21       260322.0   
2     143624       205050          Offender   2017-02-21       260322.0   
3     143624       205050          Offender   2017-02-21       260322.0   
4     143624       205050          Offender   2017-02-21       260322.0   

   STUDENT_NUMBER                INCIDENT_TITLE     ACTION_PLAN_BEGIN_DT  \
0      1149685956  Fighting/Disruptive Behavior  2017-02-22 00:00:00.000   
1      1149685956  Fighting/Disruptive Behavior                      NaN   
2      1149685956  Fighting/Disruptive Behavior                      NaN   
3      1149685956  Fighting/Disruptive Behavior  2017-02-22 00:00:00.000   
4      1149685956  Fighting/Disruptive Behavior  2017-02-28 00:00:00.000   

        ACTION_PLAN_END_DT                               ACTION_RESOLVED_DESC  \
0  2017-02-24 00:00:00.000                                                OSS   
1                      NaN                       conference setup with parent   
2                      NaN                       conference setup with parent   
3  2017-02-24 00:00:00.000                                                OSS   
4  2017-02-28 00:00:00.000  ISS  (student was suspended and returned to sc...   

   ACTION_DURATION_ASSIGNED  ACTION_DURATION_ACTUAL  ISS  OSS  REFERRAL  \
0                       3.0                     3.0    0    1         0   
1                       NaN                     NaN    0    0         1   
2                       NaN                     NaN    0    0         1   
3                       3.0                     3.0    0    1         0   
4                       1.0                     1.0    1    0         1   

   EXPULSION DISTRICT_NAME  
0          0    Cumberland  
1          0    Cumberland  
2          0    Cumberland  
3          0    Cumberland  
4          0    Cumberland

In [11]:
# Info on the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311942 entries, 0 to 311941
Data columns (total 17 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   STUDENTID                 311942 non-null  int64  
 1   INCIDENT_ID               311942 non-null  int64  
 2   INCIDENT_CATEGORY         311942 non-null  object 
 3   OFFENSE_DATE              311942 non-null  object 
 4   SCHOOL_NUMBER             311569 non-null  float64
 5   STUDENT_NUMBER            311942 non-null  int64  
 6   INCIDENT_TITLE            311859 non-null  object 
 7   ACTION_PLAN_BEGIN_DT      169603 non-null  object 
 8   ACTION_PLAN_END_DT        169505 non-null  object 
 9   ACTION_RESOLVED_DESC      93271 non-null   object 
 10  ACTION_DURATION_ASSIGNED  169201 non-null  float64
 11  ACTION_DURATION_ACTUAL    169203 non-null  float64
 12  ISS                       311942 non-null  int64  
 13  OSS                       311942 non-null  i

In [26]:
data.describe()
# not actually required, one thing we get here is on an avg the suspension time (number of days) is ~3days
# This can be better understood when take up data for a year e.g 2017 and then see the min, avg and max
# which can give some idea on continuos absentee.

STUDENTID    INCIDENT_ID  SCHOOL_NUMBER  STUDENT_NUMBER  \
count  311942.000000  311942.000000  311569.000000    3.119420e+05   
mean    97547.806240  160405.355810  260383.564716    7.607118e+08   
std     39645.370695   92566.254562      44.345429    2.123480e+09   
min     54303.000000    1000.000000  260302.000000    5.228100e+04   
25%     70619.000000   76953.250000  260357.000000    7.609470e+05   
50%     83442.000000  161442.000000  260374.000000    1.343807e+06   
75%    109751.000000  248445.750000  260424.000000    1.228741e+07   
max    248125.000000  322221.000000  260700.000000    9.989752e+09   

       ACTION_DURATION_ASSIGNED  ACTION_DURATION_ACTUAL            ISS  \
count             169201.000000           169203.000000  311942.000000   
mean                   3.175301                3.195652       0.063374   
std                    7.073659                7.082005       0.243635   
min                    0.000000                0.000000       0.000000   
25%                    1.000000                1.000000       0.000000   
50%                    2.000000                2.000000       0.000000   
75%                    3.000000                3.000000       0.000000   
max                  365.000000              365.000000       1.000000   

                 OSS       REFERRAL      EXPULSION  
count  311942.000000  311942.000000  311942.000000  
mean        0.048297       0.063076       0.000013  
std         0.214394       0.243100       0.003581  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000  
75%         0.000000       0.000000       0.000000  
max         1.000000       1.000000       1.000000

In [28]:
# Looking at each column and inferring if we want that in our final df or not.
data.columns.to_list()

['STUDENTID',
 'INCIDENT_ID',
 'INCIDENT_CATEGORY',
 'OFFENSE_DATE',
 'SCHOOL_NUMBER',
 'STUDENT_NUMBER',
 'INCIDENT_TITLE',
 'ACTION_PLAN_BEGIN_DT',
 'ACTION_PLAN_END_DT',
 'ACTION_RESOLVED_DESC',
 'ACTION_DURATION_ASSIGNED',
 'ACTION_DURATION_ACTUAL',
 'ISS',
 'OSS',
 'REFERRAL',
 'EXPULSION',
 'DISTRICT_NAME']

In [40]:
# 'INCIDENT_CATEGORY'
print(f'dType: {data.INCIDENT_CATEGORY.dtype}\n')
print(f'Categories If any? {data.INCIDENT_CATEGORY.unique()}\n')
print(f'Missing Values?: {data.INCIDENT_CATEGORY.isnull().sum()}')

dType: object

Categories If any? ['Offender']

Missing Values?: 0


The whole column is of 'Offenders' and doesn't make any difference here, but this may be needed when we merge with actual attendance df. This column can be label encoded with offenders as '1' and non offenders as '0'. This as feature in finalDF may have a good impact on model to learn that if offender the probability of being absent is high.

In [75]:
# SCHOOL_NUMBER
print(f'dType: {data.SCHOOL_NUMBER.dtype}\n')
print(f'Categories If any? \n{data.SCHOOL_NUMBER.unique()}\n')
print(f'Missing Values?: {data.SCHOOL_NUMBER.isnull().sum()}')

dType: float64

Categories If any? 
[260322. 260372. 260454. 260449. 260446. 260365. 260358. 260368. 260413.
 260383. 260401. 260455. 260369. 260326. 260408. 260440. 260318. 260321.
 260325. 260412. 260374. 260424. 260414. 260306. 260336. 260402. 260452.
 260425. 260338. 260428. 260419. 260386. 260427. 260417. 260359. 260354.
 260410. 260403. 260406. 260380. 260396. 260342. 260373. 260362. 260310.
 260459. 260430. 260415. 260418. 260367. 260352. 260404. 260426. 260363.
 260434. 260450. 260348. 260370. 260366. 260361. 260407. 260411. 260340.
 260357. 260398. 260375. 260371. 260382. 260316. 260397. 260388. 260347.
 260320. 260302. 260364. 260344. 260356. 260332. 260308. 260400. 260444.
 260312. 260448. 260405. 260432. 260416. 260700. 260456.     nan]

Missing Values?: 373


In [77]:
# INCIDENT_TITLE
print(f'dType: {data.INCIDENT_TITLE.dtype}\n')
print(f'Categories If any? \n{data.INCIDENT_TITLE.unique()}\n')
print(f'Num of Categories \n{data.INCIDENT_TITLE.nunique()}\n')
print(f'Missing Values?: {data.INCIDENT_TITLE.isnull().sum()}')

dType: object

Categories If any? 
['Fighting/Disruptive Behavior' 'Bus Misbehavior'
 'Insubordination | Fighting' ...
 'Class II ( Disruptive Behavior & Disorderly Conduct) Class I '
 '1 ISS Day' 'Inappropriate Language/ Disrespect, Disruptive Behavior w ']

Num of Categories 
18763

Missing Values?: 83


The type of Incident Title may direct to number of days suspended, which can lead to num of days absent.
We may take a look at INCIDENT_ID and INCIDENT_TITLE, if ID can give a insight on number of days suspended then we can skip the title which (actually/may) correspond to respective ID

In [90]:
# INCIDENT_ID
print(f'dType: {data.INCIDENT_ID.dtype}\n')
print(f'Categories If any? \n{data.INCIDENT_ID.unique()}\n')
print(f'Num of Categories \n{data.INCIDENT_ID.nunique()}\n')
print(f'value counts: \n{data.INCIDENT_ID.value_counts()}\n')
print(f'Missing Values?: {data.INCIDENT_ID.isnull().sum()}')

dType: int64

Categories If any? 
[205050 202841 204331 ... 205729 205025 205033]

Num of Categories 
194279

value counts: 
85838     361
148431    288
285742    279
143809    196
139068    169
         ... 
16543       1
16551       1
16582       1
16584       1
205033      1
Name: INCIDENT_ID, Length: 194279, dtype: int64

Missing Values?: 0


There are many categories of Incidents with ID '85838' being max which is Bus Misbehaviour, but number of suspension days ranges from 1 to 30. So this might not be a good idea get insight on Incident ID and Title. We can actually skip these cols

In [105]:
# ACTION_PLAN_BEGIN_DT
print(f'dType: {data.ACTION_PLAN_BEGIN_DT.dtype}\n')
print(f'Categories If any? \n{data.ACTION_PLAN_BEGIN_DT.unique()}\n')
print(f'Num of Categories \n{data.ACTION_PLAN_BEGIN_DT.nunique()}\n')
print(f'Missing Values?: {data.ACTION_PLAN_BEGIN_DT.isnull().sum()}')

dType: object

Categories If any? 
['2017-02-22 00:00:00.000' nan '2017-02-28 00:00:00.000' ...
 '2016-10-11 00:00:00.000' '2016-10-12 00:00:00.000'
 '2017-01-30 00:00:00.000']

Num of Categories 
3120

Missing Values?: 142339


In [106]:
# ACTION_PLAN_END_DT
print(f'dType: {data.ACTION_PLAN_END_DT.dtype}\n')
print(f'Categories If any? \n{data.ACTION_PLAN_END_DT.unique()}\n')
print(f'Num of Categories \n{data.ACTION_PLAN_END_DT.nunique()}\n')
print(f'Missing Values?: {data.ACTION_PLAN_END_DT.isnull().sum()}')

dType: object

Categories If any? 
['2017-02-24 00:00:00.000' nan '2017-02-28 00:00:00.000' ...
 '2016-10-13 00:00:00.000' '2016-10-14 00:00:00.000'
 '2017-01-30 00:00:00.000']

Num of Categories 
3144

Missing Values?: 142437


Both the above cols Action plan begin and end dates have lots of missing cols and can be dropped, the data here is also just the occurance date and end date. If we subtract we can get num of days suspended(absent), but since it has lot of NaN's and for now can be dropped.

In [107]:
# ACTION_RESOLVED_DESC
print(f'dType: {data.ACTION_RESOLVED_DESC.dtype}\n')
print(f'Categories If any? \n{data.ACTION_RESOLVED_DESC.unique()}\n')
print(f'Num of Categories \n{data.ACTION_RESOLVED_DESC.nunique()}\n')
print(f'value counts: \n{data.ACTION_RESOLVED_DESC.value_counts()}\n')
print(f'Missing Values?: {data.ACTION_RESOLVED_DESC.isnull().sum()}')

dType: object

Categories If any? 
['OSS' 'conference setup with parent'
 'ISS  (student was suspended and returned to school with a parent conference)'
 ... '3 days suspended OSS.'
 '3 days of ISS for 2nd period only: 2/24, 2/27, 2/28'
 'HJearing set with parent']

Num of Categories 
26053

value counts: 
OSS                                                                                                                 6430
ISS                                                                                                                 3666
2 days ISS                                                                                                          1834
Student was coached on their behavior and improvement moving forward.                                               1761
Suspended                                                                                                           1699
                                                                                       

This Col basically gives the description of the Incident, Has more than half NaN's and when we refine down there are 3 types of desc [ISS, OSS, Reference]. These 3 are present as separate cols in the dataset(just like ont hot encodings of the 3 desc), so If desc is not important then we can drop this col and use the next 3 cols.

In [135]:
# ISS
print(f'dType: {data.ISS.dtype}\n')
print(f'Categories If any? \n{data.ISS.unique()}\n')
print(f'Num of Categories \n{data.ISS.nunique()}\n')
print(f'value counts: \n{data.ISS.value_counts()}\n')
print(f'Missing Values?: {data.ISS.isnull().sum()}')

dType: int64

Categories If any? 
[0 1]

Num of Categories 
2

value counts: 
0    292173
1     19769
Name: ISS, dtype: int64

Missing Values?: 0


In [134]:
# OSS
print(f'dType: {data.OSS.dtype}\n')
print(f'Categories If any? \n{data.OSS.unique()}\n')
print(f'Num of Categories \n{data.OSS.nunique()}\n')
print(f'value counts: \n{data.OSS.value_counts()}\n')
print(f'Missing Values?: {data.OSS.isnull().sum()}')

dType: int64

Categories If any? 
[1 0]

Num of Categories 
2

value counts: 
0    296876
1     15066
Name: OSS, dtype: int64

Missing Values?: 0


In [124]:
# EXPULSION
print(f'dType: {data.EXPULSION.dtype}\n')
print(f'Categories If any? \n{data.EXPULSION.unique()}\n')
print(f'Num of Categories \n{data.EXPULSION.nunique()}\n')
print(f'value counts: \n{data.EXPULSION.value_counts()}\n')
print(f'Missing Values?: {data.EXPULSION.isnull().sum()}')

dType: int64

Categories If any? 
[0 1]

Num of Categories 
2

value counts: 
0    311938
1         4
Name: EXPULSION, dtype: int64

Missing Values?: 0


In [127]:
np.where(data.EXPULSION == 1)

(array([  6476,  35001, 103694, 269026], dtype=int64),)

In [133]:
data.iloc[269026]

STUDENTID                                                               68537
INCIDENT_ID                                                            157379
INCIDENT_CATEGORY                                                    Offender
OFFENSE_DATE                                                       2016-02-19
SCHOOL_NUMBER                                                        260449.0
STUDENT_NUMBER                                                         305239
INCIDENT_TITLE              Class III - Assault on School Personnel / Comm...
ACTION_PLAN_BEGIN_DT                                  2016-02-19 00:00:00.000
ACTION_PLAN_END_DT                                    2016-06-08 00:00:00.000
ACTION_RESOLVED_DESC        5 Day OSS ( 2/19 - 2/25). Mr. Pinkney will spe...
ACTION_DURATION_ASSIGNED                                                 72.0
ACTION_DURATION_ACTUAL                                                   72.0
ISS                                                             

In [138]:
# ACTION_DURATION_ASSIGNED 
print(f'dType: {data.ACTION_DURATION_ASSIGNED .dtype}\n')
# print(f'Categories If any? \n{data.ACTION_DURATION_ASSIGNED .unique()}\n')
print(f'Num of Categories \n{data.ACTION_DURATION_ASSIGNED .nunique()}\n')
print(f'value counts: \n{data.ACTION_DURATION_ASSIGNED .value_counts()}\n')
print(f'Missing Values?: {data.ACTION_DURATION_ASSIGNED .isnull().sum()}')

dType: float64

Num of Categories 
178

value counts: 
2.0      59409
3.0      40759
1.0      37821
5.0      10304
10.0      7454
         ...  
1.1          1
155.0        1
80.0         1
121.0        1
168.0        1
Name: ACTION_DURATION_ASSIGNED, Length: 178, dtype: int64

Missing Values?: 142741


In [137]:
# ACTION_DURATION_ACTUAL
print(f'dType: {data.ACTION_DURATION_ACTUAL.dtype}\n')
# print(f'Categories If any? \n{data.ACTION_DURATION_ACTUAL.unique()}\n')
print(f'Num of Categories \n{data.ACTION_DURATION_ACTUAL.nunique()}\n')
print(f'value counts: \n{data.ACTION_DURATION_ACTUAL.value_counts()}\n')
print(f'Missing Values?: {data.ACTION_DURATION_ACTUAL.isnull().sum()}')

dType: float64

Num of Categories 
178

value counts: 
2.0      59003
3.0      40554
1.0      37795
5.0      10500
10.0      7684
         ...  
180.0        1
82.0         1
1.1          1
147.0        1
168.0        1
Name: ACTION_DURATION_ACTUAL, Length: 178, dtype: int64

Missing Values?: 142739


The above two cols ACTION_DURATION_ASSIGNED and ACTION_DURATION_ACTUAL, gives info on number of days suspended and actual number of days suspension was served. The actual suspension gives a good idea on number of days absent when an incident is reported. So instead of using Assigned we can consider the ACTION_DURATION_ACTUAL col which gives number of days absent. But has almost ~50% NaNs.

### Cols In dataset
'STUDENTID', 'INCIDENT_ID', 'INCIDENT_CATEGORY', 'OFFENSE_DATE', 'SCHOOL_NUMBER', 'STUDENT_NUMBER', 'INCIDENT_TITLE', 'ACTION_PLAN_BEGIN_DT', 'ACTION_PLAN_END_DT', 'ACTION_RESOLVED_DESC', 'ACTION_DURATION_ASSIGNED', 'ACTION_DURATION_ACTUAL', 'ISS', 'OSS', 'REFERRAL', 'EXPULSION', 'DISTRICT_NAME'

### Cols needed for further analysis
'STUDENTID', 'OFFENSE_DATE', 'ACTION_DURATION_ACTUAL', 'ISS', 'OSS', 'REFERRAL'

### Cols can be dropped?
'INCIDENT_ID', 'INCIDENT_CATEGORY', 'SCHOOL_NUMBER', 'STUDENT_NUMBER', 'INCIDENT_TITLE', 'ACTION_PLAN_BEGIN_DT', 'ACTION_PLAN_END_DT', 'ACTION_RESOLVED_DESC', 'ACTION_DURATION_ASSIGNED', 'EXPULSION', 'DISTRICT_NAME'

In [149]:
# New df for further analysis
disc_df = data[['STUDENTID', 'OFFENSE_DATE', 'ACTION_DURATION_ACTUAL', 'ISS', 'OSS', 'REFERRAL']]
print(f'Shape of dataset: {disc_df.shape}\n')
print(f'Missing values in dataset: \n{disc_df.isnull().sum()}\n')
disc_df.head(10)

Shape of dataset: (311942, 6)

Missing values in dataset: 
STUDENTID                      0
OFFENSE_DATE                   0
ACTION_DURATION_ACTUAL    142739
ISS                            0
OSS                            0
REFERRAL                       0
dtype: int64



STUDENTID OFFENSE_DATE  ACTION_DURATION_ACTUAL  ISS  OSS  REFERRAL
0     143624   2017-02-21                     3.0    0    1         0
1     143624   2017-02-21                     NaN    0    0         1
2     143624   2017-02-21                     NaN    0    0         1
3     143624   2017-02-21                     3.0    0    1         0
4     143624   2017-02-21                     1.0    1    0         1
5      68649   2017-02-21                     3.0    0    1         0
6      68649   2017-02-21                     NaN    0    0         1
7      68649   2017-02-21                     NaN    0    0         1
8      68649   2017-02-21                     3.0    0    1         0
9      68649   2017-02-21                     1.0    1    0         1

In [152]:
# Save this df to csv
disc_df.to_csv('new_discipline.csv', sep=',', index=False)